content/blog/2022/2022-07-10-jackson.md

```diff
@@ -0,0 +1,547 @@
+Annotations mainly affect how the data is read, written or even both. Let's explore some of those annotations based on their categories.
+
+### Read Annotations
+They affect how Jackson converts JSON into POJOs.
+
+### @JsonSetter
+This is useful when we want to match a field in the JSON string to a field in the POJO where their names don't match.
+```java
+@NoArgsConstructor  
+@AllArgsConstructor  
+@Getter  
+public class Car {  
+    @JsonSetter("carBrand")  
+    private String brand;  
+}
+```
+```json
+{  
+  "carBrand" : "BMW"  
+}
+```
+```java
+public class JacksonTest {
+	...
+	@Test  
+	void fileToCar() throws IOException {  
+	    File file = new File("src/test/resources/car.json");  
+	  
+	    Car car = objectMapper.readValue(file, Car.class);  
+	  
+	    assertThat(car.getBrand()).isEqualTo("BMW");  
+	}
+}
+```
+
+### @JsonAnySetter
+This annotation is useful for cases where the JSON contains some fields that are not declared in the POJO. It is used with a setter method that is called for every unrecognized field.
+
+```java
+public class Car {  
+    @JsonSetter("carBrand")  
+    private String brand;  
+    private Map<String, String> unrecognizedFields = new HashMap<>();  
+  
+    @JsonAnySetter  
+  public void allSetter(String fieldName, String fieldValue) {  
+        unrecognizedFields.put(fieldName, fieldValue);  
+    }  
+}
+```
+`carUnrecognized.json` file:
+```json
+{  
+  "carBrand" : "BMW",  
+  "productionYear": 1996  
+}
+```
+
+```java
+public class JacksonTest {
+	...
+	@Test  
+	void fileToUnrecognizedCar() throws IOException {  
+	    File file = new File("src/test/resources/carUnrecognized.json");  
+	  
+	    Car car = objectMapper.readValue(file, Car.class);  
+	  
+	    assertThat(car.getUnrecognizedFields()).containsKey("productionYear");  
+	}
+}
+```
+
+### Write Annotations
+They affect how Jackson converts POJOs into JSON.
+
+### @JsonGetter
+This is useful when we want to map a POJOs field to a JSON field using a different name.
+For example, say we have this Cat class with the field `name`, but we want its Json name to be `catName`.
+```java
+@NoArgsConstructor  
+@AllArgsConstructor  
+public class Cat {  
+    private String name;  
+  
+    @JsonGetter("catName")  
+    public String getName() {  
+        return name;  
+    }  
+}
+```
+
+```java
+public class JacksonTest {
+	...
+	@Test  
+	void catToJson() throws JsonProcessingException {  
+	    Cat cat = new Cat("Monica");  
+	  
+	    String json = objectMapper.writeValueAsString(cat);  
+	  
+	    System.out.println(json);  
+	}
+}
+```
+This will output
+```json
+{
+	"catName":"Monica"
+}
+```
+
+### @JsonAnyGetter
+This annotation allows us to treat a Map object as a source of JSON properties.
+Say we have this map as a filed in a POJO
+```java
+Map<String, String> map = Map.of(
+	"name", "Jack", 
+	"surname", "wolfskin"
+);
+```
+Jackson will serialize it to
+
+```json
+{
+	"name":"Jack",
+	"surname":"wolfskin"
+}
+```
```

Comment:
The code example for `@JsonAnyGetter` looks incomplete, can you fix that?

Reply:
Just did